File: lc-blind75/ProductOfArrayExceptSelf(lc-238)/Better/ProductSelf.java
import java.util.Scanner;

public class ProductSelf {
    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        int n=sc.nextInt();
        int nums[]=new int [n];
        for(int i=0;i<n;i++){
            nums[i]=sc.nextInt();
        }
        int res[]=productExceptSelf(nums);
        for(int x:res){
            System.out.println(x);
        }
        sc.close();
    }
    public static int[] productExceptSelf(int nums[]){
        int n=nums.length;
        int res[]=new int [n];
        int pre=1;
        for(int i=0;i<n;i++){
            res[i]=pre;
            pre*=nums[i];
        }
        int post=1;
        for(int j=n-1;j>=0;j--){
            res[j]*=post;
            post*=nums[j];
        }
        return res;
    }
}

File: lc-blind75/ProductOfArrayExceptSelf(lc-238)/Better/ProductSelf.py
def productExceptSelf(nums):
    n=len(nums)
    res=[]
    pre=1
    for i in range(n):
        res.append(pre)
        pre*=nums[i]
    post=1
    for j in range(n-1,-1,-1):
        res[j]*=post
        post*=nums[j]
    return res


n=int(input())
nums=list(map(int,input().split()))
print(productExceptSelf(nums))